Index: darkstar-server/src/main/java/com/sun/sgs/impl/kernel/Kernel.java
===================================================================
--- darkstar-server/src/main/java/com/sun/sgs/impl/kernel/Kernel.java	(revision 82)
+++ darkstar-server/src/main/java/com/sun/sgs/impl/kernel/Kernel.java	(revision 83)
@@ -21,48 +21,35 @@
 
 import com.sun.sgs.app.AppListener;
 import com.sun.sgs.app.NameNotBoundException;
-
 import com.sun.sgs.auth.Identity;
 import com.sun.sgs.auth.IdentityAuthenticator;
 import com.sun.sgs.auth.IdentityCoordinator;
-
 import com.sun.sgs.impl.auth.IdentityImpl;
-
 import com.sun.sgs.impl.kernel.StandardProperties.StandardService;
-
 import com.sun.sgs.impl.profile.ProfileCollectorImpl;
 import com.sun.sgs.impl.profile.ProfileRegistrarImpl;
-
 import com.sun.sgs.impl.service.data.DataServiceImpl;
-
 import com.sun.sgs.impl.service.transaction.TransactionCoordinator;
 import com.sun.sgs.impl.service.transaction.TransactionCoordinatorImpl;
-
 import com.sun.sgs.impl.sharedutil.LoggerWrapper;
-
 import com.sun.sgs.impl.util.AbstractKernelRunnable;
 import com.sun.sgs.impl.util.Version;
-
 import com.sun.sgs.kernel.ComponentRegistry;
-
 import com.sun.sgs.profile.ProfileCollector;
 import com.sun.sgs.profile.ProfileListener;
 import com.sun.sgs.profile.ProfileProducer;
 import com.sun.sgs.profile.ProfileRegistrar;
-
 import com.sun.sgs.service.DataService;
 import com.sun.sgs.service.Service;
 import com.sun.sgs.service.TransactionProxy;
 
+import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
-
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
-
 import java.util.ArrayList;
 import java.util.Properties;
-
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
@@ -851,7 +838,12 @@
         
         // check the standard properties
         checkProperties(appProperties, args[0]);
-        
+
+        // [darkstar-exp] interpret app root as relative to the properties file
+        File propertiesDir = new File(args[0]).getParentFile();
+        File appRoot = new File(propertiesDir, appProperties.getProperty(StandardProperties.APP_ROOT));
+        appProperties.setProperty(StandardProperties.APP_ROOT, appRoot.getPath());
+
         // boot the kernel
         new Kernel(appProperties);
     }
